storage/src/binary/mod.rs

```diff
@@ -0,0 +1,457 @@
+/// of [commonware_cryptography::Digest].
+#[derive(Clone, Debug)]
+pub struct Tree<H: Hasher> {
+    /// Levels of the tree: level 0 contains the leaves, and each subsequent level
+    /// contains the parent nodes computed from the previous level.
+    ///
+    /// The last level contains a single node: the root.
+    levels: Vec<Vec<H::Digest>>,
+}
+
+impl<H: Hasher> Tree<H> {
+    /// Builds a Merkle Tree from a slice of leaf digests.
+    ///
+    /// If `leaves` is empty, returns `None`.
+    ///
+    /// # Warning
+    ///
+    /// It is not safe to insert `H::Digest::default()` as a leaf. This could lead to
+    /// a proof being generated for a non-existent leaf.
+    pub fn new(hasher: &mut H, leaves: Vec<H::Digest>) -> Option<Self> {
+        // Ensure there are non-zero leaves.
+        if leaves.is_empty() {
+            return None;
+        }
+
+        // Level 0: the leaves.
+        let mut levels = Vec::new();
+        levels.push(leaves);
+
+        // Build higher levels until we reach the root.
+        let default = H::Digest::default();
+        while levels.last().unwrap().len() > 1 {
+            let current_level = levels.last().unwrap();
+            let mut next_level = Vec::with_capacity((current_level.len() + 1) / 2);
+            for chunk in current_level.chunks(2) {
+                let left = &chunk[0];
+                let right = if chunk.len() == 2 {
+                    &chunk[1]
+                } else {
+                    // Use the default digest for the right child if no right child exists.
+                    &default
+                };
+                next_level.push(combine(hasher, left, right));
+            }
+            levels.push(next_level);
+        }
+        Some(Self { levels })
+    }
+
+    /// Returns a reference to the Merkle root, if the tree is non-empty.
+    pub fn root(&self) -> H::Digest {
+        self.levels.last().unwrap().first().unwrap().clone()
+    }
+
+    /// Generates a Merkle proof for the leaf at `leaf_index`.
+    ///
+    /// The proof contains the sibling hash at each level needed to reconstruct the root.
+    pub fn prove(&self, leaf_index: usize) -> Option<Proof<H>> {
+        if leaf_index >= self.levels[0].len() {
+            return None;
+        }
+        let mut proof_hashes = Vec::new();
+        let mut index = leaf_index;
+        // For each level (except the root level) record the sibling.
+        for level in &self.levels {
+            if level.len() == 1 {
+                break; // Reached the root.
+            }
+            let sibling_index = if index % 2 == 0 { index + 1 } else { index - 1 };
+            let sibling = if sibling_index < level.len() {
+                level[sibling_index].clone()
+            } else {
+                // Use the default digest for the right child if no right child exists.
+                H::Digest::default()
+            };
+            proof_hashes.push(sibling);
+            index /= 2;
```

Comment:
This will continue with index == 0 potentially muiltiple times if there is "extra" proof data, but this is a good thing as it should prevent proof malleability by taking on extra stuff onto the proof.  Can you add a test case for that?  Just tack on an extra hash to a valid proof and make sure its verification fails.

Reply:
Good call 👍